Plasorak/fixes v3.1.0

plasorak opened the pull request:
This PR implements changes requested for v3.1.1 patch releases, namely:
 - Fixes enable/disable
 - Move error state outside the FSM, so that we can recover errored parent nodes
 - Notify GUI user when an application crashes
 - Makes FELIX work for the card controller and ruflx
 - Fix thread pinning on k8spm

bieryAtFnal committed:
Merge pull request #117 from DUNE-DAQ/prep-release/dunedaq-v3.1.0

Merging the prep-release/dunedaq-v3.1.0 branch to develop now that the release is being finalized

plasorak committed:
massive hack to make flxcard app work

plasorak committed:
execute_script for thread pinning

plasorak committed:
fix the api name

plasorak committed:
fix include/exclude

plasorak committed:
always terminate

plasorak committed:
make the Ack renderable

plasorak committed:
better can_execute

plasorak committed:
make thread pinning work for k8s

plasorak committed:
more info

plasorak committed:
fix node path in the gui

plasorak committed:
better GUI

plasorak committed:
make the error state outside the FSM, so that we can recover

plasorak committed:
make error visible on all the nodes

plasorak committed:
typo

plasorak committed:
make sure the return code is an int

plasorak committed:
make sure the return code is an int

plasorak committed:
stop when the command sequence isn't right

plasorak committed:
make sure forced commands are executed only when needed

plasorak committed:
fail quickly when the command make the app crash

plasorak committed:
don't stop when the command fails

plasorak committed:
rm dirty hack

plasorak committed:
allows for terminate even if the child node is none

plasorak committed:
proper use of felix resources

plasorak committed:
easthethics

plasorak committed:
fix typo for thread pinning

JonathanHancock0 reviewed (approved):
The code has been demonstrated to work correctly when using the felix emulator in kubernetes.

plasorak commented:
To be more specific, here are the changes that went in:

<h2>Summary of changes</h2>

<h3>Include/Exclude commands</h3>

 - `terminate` now always terminate excluded nodes.
 - Fixed a bug in which an excluded node still received commands (`list.remove(app)` was executed while looping over `list`)

<h3>Error states</h3>

 - This used to be an FSM state. However it was impossible to exit it (simply because the nodes doesn't know what state it should be in after exiting). So I removed it from the FSM, and made "error" a boolean node member. That way the error state can be "toggled", and a node can exit the error state and still know what was its state.
 - I added some logic to come in and out of error state based on included and excluded nodes (`resolve_error` functions).

<h3>Command sending check</h3>

I've added more complex logic to understand whether a command can be executed or not. This is needed because there are several reasons why a command should or shouldn't be executed on a node:
 - Errored nodes:
   - Mustn't block the terminate command
   - Must block the execution of any other command, other than include/exclude 
 - Excluded nodes:
   - Must still be terminated
   - Must only respond to the "include" command
   - Must not propagate their error state upwards (to parent nodes)

`--force`'d commands will check whether the command can be executed, but only execute on valid FSM transition. So they will be sent to included errored and healthy nodes.

Additionnally there are some complixity on composite commands (`shutdown` for example, execute _optionally_ `disable_triggers`, `drain_dataflow`, `stop_trigger_sources`, `stop`, `scrap`, and _always_ execute `terminate`). Here is the way it behaves:
 - When there isn't any problem in the tree, the composite command will start when from the state of the FSM (say the topnode is `dataflow_drained`, `shutdown` will `stop_trigger_sources`, `stop` etc.). If one of the command isn't optional, the composite command will stop (so for example topnode is `configured`, and you are trying to `stop_run` - `drain_dataflow` can't be executed so it should fail)
 - When `--force` is passed to a composite command, nanorc should still check that the FSM transition is possible, and send (optional) commands to all the included nodes (whether in error or not).

<h3>GUI</h3>

 - The problem was that it wouldn't say when an app would crash, this was a bug in the GUI.
 - The GUI now produces a browser alert when an app is crashed.
 - Thread pinning command wasn't visible on the GUI due to a typo in the REST API.
 - Corrected a bug for the commands that had "node_path" parameter (that was causing the `exclude` cmd to not work on the GUI).

<h3>FELIX</h3>

Now works with 2 types of resources:
 - flx0-data for readout application using felix
 - flx0-ctrl for felix card controller

<h3>Thread pinning</h3>

Added the missing function in k8spm.


<h2>Testing</h2>

Following tests where done with and without the GUI (`--web` toggle), on both `nanorc` and `nano04rc` and with/without `k8s`:
 - I can run with felix card controller and ruflx in emulator mode on `np04-srv-028`.
 - I can exclude a node and go through the FSM states, using composite commands or not, even if the node is not in error.
 - I can go through the FSM states after an application went into error and was excluded.
 - I get a notification on the GUI if I manually kill an application (only for `sshpm`, no k8s), and the app table gets updated accordingly.
 - After killing an application I can exclude it and move through the FSM states and achieve a clean shutdown.
 - If an application crashes on a command (for example, sending configure to a felixcardctrl app on a server that has no FELIX), I can exclude that application, and continue through the FSM states.
 - I can include nodes that were excluded.
